**CUISINE CLASSIFICATION - TASK 3**

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving Dataset .csv to Dataset .csv


1) Loading dataset and importing libraries

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

res_data = pd.read_csv('Dataset .csv')
print(res_data.head())

   Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM 

In [ ]:
res_data.columns

Index(['Restaurant ID', 'Restaurant Name', 'Country Code', 'City', 'Address',
       'Locality', 'Locality Verbose', 'Longitude', 'Latitude', 'Cuisines',
       'Average Cost for two', 'Currency', 'Has Table booking',
       'Has Online delivery', 'Is delivering now', 'Switch to order menu',
       'Price range', 'Aggregate rating', 'Rating color', 'Rating text',
       'Votes'],
      dtype='object')

2)data cleaning

In [ ]:
res_data=res_data.drop(columns=['Restaurant ID','Country Code','City','Address','Locality','Locality Verbose','Longitude','Latitude','Average Cost for two','Currency','Has Table booking','Has Online delivery','Is delivering now','Switch to order menu','Price range','Aggregate rating','Rating color','Rating text','Votes'])
res_data

Restaurant Name                          Cuisines
0             Le Petit Souffle        French, Japanese, Desserts
1             Izakaya Kikufuji                          Japanese
2       Heat - Edsa Shangri-La  Seafood, Asian, Filipino, Indian
3                         Ooma                   Japanese, Sushi
4                  Sambo Kojin                  Japanese, Korean
...                        ...                               ...
9546               Naml۱ Gurme                           Turkish
9547              Ceviz A��ac۱   World Cuisine, Patisserie, Cafe
9548                     Huqqa            Italian, World Cuisine
9549               A���k Kahve                   Restaurant Cafe
9550  Walter's Coffee Roastery                              Cafe

[9551 rows x 2 columns]

In [ ]:
#handling missing values
res_data.isnull().sum()

Restaurant Name    0
Cuisines           9
dtype: int64

In [ ]:
res_data = res_data.dropna(subset=['Cuisines'])

3)Data Preprocessing

In [ ]:
#Dealing with multiple values in cuisines
res_data['Cuisines'] = res_data['Cuisines'].str.split(', ')
res_data=res_data.explode('Cuisines')
res_data.head()

<ipython-input-80-6be0ff1df6ce>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_data['Cuisines'] = res_data['Cuisines'].str.split(', ')


Restaurant Name  Cuisines
0        Le Petit Souffle    French
0        Le Petit Souffle  Japanese
0        Le Petit Souffle  Desserts
1        Izakaya Kikufuji  Japanese
2  Heat - Edsa Shangri-La   Seafood

4)Data Spliting

In [ ]:
# Encode the 'Cuisines' labels
le = LabelEncoder()
res_data['Cuisines'] = le.fit_transform(res_data['Cuisines'])

# Prepare features using TfidfVectorizer on 'Restaurant Name'
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(res_data['Restaurant Name'])

# Target variable
y = res_data['Cuisines']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=40)

5)Model Training and observation

**1) LOGISTIC REGRESSION MODEL**

In [ ]:
# Initialize and train the Logistic Regression model
lr_model = LogisticRegression(max_iter=1000)
lr_model.fit(X_train, y_train)

# Make predictions with Logistic Regression
lr_y_pred = lr_model.predict(X_test)

# Evaluate Logistic Regression model
lr_accuracy = accuracy_score(y_test, lr_y_pred)
print(f'Logistic Regression Accuracy: {lr_accuracy:.2f}')
print('Logistic Regression Confusion Matrix:')
print(confusion_matrix(y_test, lr_y_pred))

Logistic Regression Accuracy: 0.24
Logistic Regression Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 3 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


**2) RANDOM FOREST CLASSIFIER MODEL**

In [ ]:
# Initialize and train the Random Forest model
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions with Random Forest
rf_y_pred = rf_model.predict(X_test)

# Evaluate Random Forest model
rf_accuracy = accuracy_score(y_test, rf_y_pred)
print(f'Random Forest Accuracy: {rf_accuracy:.2f}')
print('Random Forest Confusion Matrix:')
print(confusion_matrix(y_test, rf_y_pred))

Random Forest Accuracy: 0.16
Random Forest Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 2 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


**3)SUPPORT VECTOR MACHINE MODEL**

In [ ]:
# Initialize and train the Support Vector Machine model
svm_model = SVC(kernel='linear', random_state=40)
svm_model.fit(X_train, y_train)

# Make predictions with Support Vector Machine
svm_y_pred = svm_model.predict(X_test)

# Evaluate Support Vector Machine model
svm_accuracy = accuracy_score(y_test, svm_y_pred)
print(f'Support Vector Machine Accuracy: {svm_accuracy:.2f}')
print('Support Vector Machine Confusion Matrix:')
print(confusion_matrix(y_test, svm_y_pred))

Support Vector Machine Accuracy: 0.25
Support Vector Machine Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 6 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


**4) DECISION TREE CLASSIFIER MODEL**

In [ ]:
# Initialize and train the Decision Tree model
dt_model = DecisionTreeClassifier(random_state=40)
dt_model.fit(X_train, y_train)

# Make predictions with Decision Tree
dt_y_pred = dt_model.predict(X_test)

# Evaluate Decision Tree model
dt_accuracy = accuracy_score(y_test, dt_y_pred)
print(f'Decision Tree Accuracy: {dt_accuracy:.2f}')
print('Decision Tree Confusion Matrix:')
print(confusion_matrix(y_test, dt_y_pred))

Decision Tree Accuracy: 0.14
Decision Tree Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 1 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


**CONCLUSION :

The best model is SVM with the highest accurace of 0.25, followed by Logistic Regression having 0.24 accuracy.

Decision Tree Classifier (Accuracy : 0.14) and Random Forest Classifier (Accuracy : 0.16) performed poorly for the purpose**